File: client/lib/crypt/hkdf.ex
defmodule Crypt.Hkdf do
  @moduledoc """
  Documentation for `Hkdf`.
  """

  @doc """
  Derives a key from an input key material.

  ## Examples

      iex> input = Base.decode16!("784587B71309D1C4774F6FDF9FE5160753C40EF67F145CA62177C6CA36C2151D")
      iex> length = 32
      iex> salt = Base.decode16!("48EA16CCF2829D493F9ADBADE344F061")
      iex> info = "info"
      iex> key = Crypt.Hkdf.derive(input, length, salt, info)
      key
  """

  def derive(input, length, salt \\ "", info \\ "") do
    prk = extract(input, salt)
    okm = expand(prk, info, length)

    okm
  end

  defp extract(input, salt) do
    prk = :crypto.mac(:hmac, :sha512, salt, input)

    prk
  end

  defp expand(prk, info, length) do
    hash_length = 64

    n = Float.ceil(length / hash_length) |> round()

    full =
      Enum.scan(1..n, "", fn index, prev ->
        data = prev <> info <> <<index>>
        :crypto.mac(:hmac, :sha512, prk, data)
      end)
      |> Enum.reduce("", &Kernel.<>(&2, &1))

    <<output::unit(8)-size(length), _::binary>> = full
    <<output::unit(8)-size(length)>>
  end
end

File: client-go/internal/contact/ratchet/hkdf.go
package ratchet

import (
	"crypto/hmac"
	"crypto/sha512"
	"io"
)

// Derive generates a key from input key material using HKDF.
func derive(input, salt, info []byte, length int) ([]byte, error) {
	prk := extract(input, salt)

	okm, err := expand(prk, info, length)
	if err != nil {
		return nil, err
	}

	return okm, nil
}

func extract(input, salt []byte) []byte {
	if salt == nil {
		salt = make([]byte, sha512.Size)
	}

	h := hmac.New(sha512.New, salt)
	h.Write(input)
	return h.Sum(nil)
}

func expand(prk, info []byte, length int) ([]byte, error) {
	hashLength := sha512.Size
	n := (length + hashLength - 1) / hashLength

	var okm []byte
	var previousBlock []byte

	for i := 1; i <= n; i++ {
		h := hmac.New(sha512.New, prk)
		h.Write(previousBlock)
		h.Write(info)
		h.Write([]byte{byte(i)})
		previousBlock = h.Sum(nil)
		okm = append(okm, previousBlock...)
	}

	if len(okm) < length {
		return nil, io.ErrShortBuffer
	}

	return okm[:length], nil
}
